### PR TITLE
timer counter and tracker implementation

### PR DESCRIPTION
**Pull request by xavluiz**
this implements the time counter logic and tracker

Screen Shot 2020-07-29 at 6 18 19 AM

### CODE REVIEW

File: lib/UtilManager.js [hunk 1]
```diff
@@ -220,7 +230,17 @@ utilMgr.isMac = () => {
...
 
+utilMgr.getJwt = () => {
+  if (jwt === null) {
+    jwt = utilMgr.getItem("jwt");
+  }
+  return jwt;
+}
+
 utilMgr.setItem = (key, value) => {
+  if (key === "jwt") {
```
Comment (bjacobson26) [inline lib/UtilManager.js:241]: maybe put this after 244 to ensure it successfully saves

File: lib/UtilManager.js [hunk 2]
```diff
@@ -937,4 +940,19 @@ utilMgr.getLatestPayload = () => {
     return latestPayload;
 };
 
+utilMgr.coalesceNumber = (val, defaultVal = 0) => {
+    if (val === null || val === undefined || isNaN(val)) {
```
Comment (bjacobson26) [inline lib/UtilManager.js:944]:
```
return val || defaultVal
```

?

File: lib/managers/ProjectManager.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            }
+        }
+      }
+  }
+
+  return project;
+};
+
+projectMgr.getFirstProjectDirectory = () => {
+    if (
```
Comment (bjacobson26) [inline lib/managers/ProjectManager.js:87]: this one is probably safe to do:

```
return atom.workspace.project?.rootDirectories[0]?.path || ""
```

File: lib/repo/GitUtil.js [hunk 4]
```diff
@@ -344,4 +346,176 @@ gitUtil.getThisWeek = () => {
...
+  const cmd = `git log --pretty="%an" ${fileName}`;
+
+  // get the list of users that modified this file
+  let resultList = await execUtil.getCommandResultList(cmd, directory);
+  if (!resultList) {
+      // something went wrong, but don't try to parse a null or undefined str
+      return 0;
+  }
+
+  if (resultList.length > 0) {
```
Comment (bjacobson26) [inline lib/repo/GitUtil.js:373]: is this just getting a count of unique elements in a list?

I think you can do:
```
var items = [4,5,4,6,3,4,5,2,23,1,4,4,4]
var uniqueItems = Array.from(new Set(items))
=> [4, 5, 6, 3, 2, 23, 1]
```

File: lib/repo/GitUtil.js [hunk 5]
```diff
@@ -344,4 +346,176 @@ gitUtil.getThisWeek = () => {
...
+
+    return resultList.length;
+};
+
+function stripOutSlashes(str) {
+  var parts = str.split("//");
+  return parts.length === 2 ? parts[1] : str;
+}
+
+function stripOutAmpersand(str) {
```
Comment (bjacobson26) [inline lib/repo/GitUtil.js:462]: should name this `stripeOutAtSign`
ampersand is `&`

File: lib/UtilManager.js (hunk 1 above)
Comment (xavluiz) [reply lib/UtilManager.js:227]: sorry i didn't get back to this request. based on the instrumentation of fetching from the file and ensuring we don't have a per window caching issue, we'll just fetch from the file and i'll remove this logic.

File: lib/UtilManager.js (hunk 2 above)
Comment (xavluiz) [reply lib/UtilManager.js:930]: yeah, lets go with that. good idea

File: lib/managers/ProjectManager.js (hunk 3 above)
Comment (xavluiz) [reply lib/managers/ProjectManager.js:87]: i'll have to update babel to interpret conditional chaining

File: lib/repo/GitUtil.js (hunk 4 above)
Comment (xavluiz) [reply lib/repo/GitUtil.js]: yeah, i like that, cleaner. i've updated it

File: lib/repo/GitUtil.js (hunk 5 above)
Comment (xavluiz) [reply lib/repo/GitUtil.js:455]: lol, yes. updating

Comment (xavluiz) [reply lib/repo/GitUtil.js]: updated

**Comment by xavluiz** [comment]
I'm going to merge this PR as it's become hefty, but will test the interactions the rest of the day. I've also merged in the logic to capture the charsPasted.

**Pull request merged**